lnwallet/musig_session.go

```diff
@@ -0,0 +1,535 @@
+	}
+}
+
+// FinalizeSession finalizes the session given the signer nonce (remote nonce).
+// This is called before signing or verifying a new commitment.
+//
+// TODO(roasbeef): make private again, add NewMusigSessionthat calls above then
+// calls thisd
+func (m *MusigSession) FinalizeSession(signingNonce musig2.Nonces) error {
+	var (
+		localNonce, remoteNonce musig2.Nonces
+		err                     error
+	)
+
+	// First, we'll stash the freshly generated signing nonce. Depending on
+	// who's commitment we're handling, this'll either be our generated
+	// nonce, or the one we just got from the remote party.
+	m.nonces.SigningNonce = signingNonce
+
+	switch {
+
+	// If we're making a session for the remote commitment, then the nonce
+	// we use to sign is actually will be the signing nonce for the
+	// session, and their nonce the verification nonce.
+	case m.remoteCommit:
+		localNonce = m.nonces.SigningNonce
+		remoteNonce = m.nonces.VerificationNonce
+
+	// Otherwise, we're generating/receiving a signature for our local
+	// commitment (to broadcast), so now our verification nonce is the one
+	// we've already generated, and we want to bind their new signing
+	// nonce.
+	default:
+		localNonce = m.nonces.VerificationNonce
+		remoteNonce = m.nonces.SigningNonce
+	}
+
+	tweakDesc := input.MuSig2Tweaks{
+		TaprootBIP0086Tweak: true,
+	}
+	m.session, err = m.signer.MuSig2CreateSession(
+		m.localKey.KeyLocator, m.signerKeys, &tweakDesc,
+		[][musig2.PubNonceSize]byte{remoteNonce.PubNonce},
+		musig2.WithPreGeneratedNonce(&localNonce),
+	)
+	if err != nil {
+		return err
+	}
+
+	// We'll need the raw combined nonces later to be able to verify
+	// partial signatures, and also combine partial signatures, so we'll
+	// generate it now ourselves.
+	m.combinedNonce, err = musig2.AggregateNonces([][musig2.PubNonceSize]byte{
+		m.nonces.SigningNonce.PubNonce,
+		m.nonces.VerificationNonce.PubNonce,
+	})
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// taprootKeyspendSighash generates the sighash for a taproot key spend. As
+// this is a musig2 channel output, the keyspend is the only path we can take.
+func taprootKeyspendSighash(tx *wire.MsgTx, pkScript []byte,
+	value int64) ([]byte, error) {
+
+	prevOutputFetcher := txscript.NewCannedPrevOutputFetcher(
+		pkScript, value,
+	)
+
+	sigHashes := txscript.NewTxSigHashes(tx, prevOutputFetcher)
+
+	return txscript.CalcTaprootSignatureHash(
+		sigHashes, txscript.SigHashDefault, tx, 0, prevOutputFetcher,
+	)
+}
+
+// SignCommit signs the passed commitment w/ the current signing (relative
+// remote) nonce. Given nonces should only ever be used once, once the method
+// returns a new nonce is returned, w/ the existing nonce blanked out.
+func (m *MusigSession) SignCommit(tx *wire.MsgTx) (*MusigPartialSig, error) {
+	// If we already have a session, then we don't need to finalize as this
+	// was done up front (symmetric nonce case, like for co-op close).
+	if m.session == nil {
```

Comment:
should we explicitly block this path if this is for our local commit (ie, `m.remoteCommit` == false)? just so that we prevent calling FinaliseSession and then overwriting the signing nonce with what is actually the verification nonce. I mean, I dont think we would ever call `SignCommit` for our local commitment unless we have already revoked the state & ie, must have received the remote nonce & created `m.session` already. but perhaps better to be safe?

Reply:
I think we should add this check

Reply:
Makes sense, added the extra protection.

Reply:
looks like it was added but then removed?

Reply:
im thinking we should maybe just explicitly error out if `session == nil` at this point to ensure that the caller definitely first called `Finalise` first

Reply:
Ah I added it, but then realized I implemented an incorrect version. I think review would better be focused on the actual call sites here (in the state machine, etc), but can't hurt to add this extra level of protection in an assertion. 

Note that the only time we call this method for the local commitment is when we actually go to broadcast to force close. At that point, channel operations have ceased.

Tacked on a sub-test for this case as well.